### ISSUE TITLE
Tensorflow lite conversion problem

### ISSUE THREAD

**Issue by tulasiram58827** [tensorflow/tensorflow#33429]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
- TensorFlow installed from (source or binary):
- TensorFlow version (or github SHA if from source):

**Provide the text output from tflite_convert**

```
Some of the operators in the model are not supported by the standard TensorFlow Lite runtime and are not recognized by TensorFlow. If you have a custom implementation for them you can disable this error with --allow_custom_ops, or by setting allow_custom_ops=True when calling tf.lite.TFLiteConverter(). Here is a list of builtin operators you are using: FULLY_CONNECTED, RESHAPE, TRANSPOSE. Here is a list of operators for which you will need custom implementations: EmptyTensorList, TensorListFromTensor, TensorListReserve, TensorListStack, While.

```
Graph code:
self.inputs = tf.placeholder(tf.float32, [None, None, num_features], name='inputs')

        # # Here we use sparse_placeholder that will generate a
        # # SparseTensor required by ctc_loss op.
        # # https://www.tensorflow.org/api_docs/python/tf/sparse/SparseTensor
        # # https://www.tensorflow.org/api_docs/python/tf/nn/ctc_loss
        self.targets = tf.sparse_placeholder(tf.int32, name='targets')

        # # 1d array of size [batch_size]
        self.seq_len = tf.placeholder(tf.int32, [None], name='seq_len')
        self.lstm_cell = tf.lite.experimental.nn.TFLiteLSTMCell(num_hidden)
        self.outputs, _ =tf.lite.experimental.nn.dynamic_rnn(self.lstm_cell,self.inputs,dtype='float32')

        self.shape = tf.shape(self.inputs)
        self.batch_s, self.max_time_steps = self.shape[0], self.shape[1]

        # Reshaping to apply the same weights over the timesteps
        self.outputs = tf.reshape(self.outputs, [-1, num_hidden])

        # Truncated normal with mean 0 and stdev=0.1
        # Tip: Try another initialization
        # see https://www.tensorflow.org/versions/r0.9/api_docs/python/contrib.layers.html#initializers
        self.W = tf.Variable(tf.truncated_normal([num_hidden,
                                             num_classes],
                                            stddev=0.1))
        # Zero initialization
        # Tip: Is tf.zeros_initializer the same?
        self.b = tf.Variable(tf.constant(0., shape=[num_classes]))

        # Doing the affine projection
        self.logits = tf.matmul(self.outputs, self.W) + self.b

        # Reshaping back to the original shape
        self.logits = tf.reshape(self.logits, [self.batch_s, -1, num_classes])

        # Time major
        self.logits = tf.transpose(self.logits, (1, 0, 2))

        self.logits = tf.identity(self.logits,name="output")

        self.loss = tf.nn.ctc_loss(self.targets, self.logits, self.seq_len)
        self.cost = tf.reduce_mean(self.loss)

        with tf.variable_scope("gs"):
            self.global_step = tf.Variable(0, name='global_step', trainable=False)

        # optimizer = tf.train.AdamOptimizer().minimize(cost)
        # optimizer = tf.train.MomentumOptimizer(learning_rate=0.01, momentum=0.9).minimize(cost)
        self.optimizer = tf.train.AdamOptimizer(learning_rate=5e-4)
        self.gvs = self.optimizer.compute_gradients(self.cost)
        self.clipped = []
        for grad, var in self.gvs:
            grad = tf.clip_by_value(grad, -1., 1.)
            self.clipped.append((grad, var))
            self.train_op = self.optimizer.apply_gradients(self.clipped, global_step=self.global_step)

**Any other info / logs**

Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by haozha111** [comment]
Hi,

The conversion fails because it has a dynamic rnn in your model, and the ops used by the rnn are 'TensorList' ops and while op, which is not supported by TOCO converter.

We are currently working on a new converter that could handle all these use cases. Please stay tuned!

**Comment by tulasiram58827** [comment]
But I am using tf.lite.experimental.dynamic_rnn which is supported as TFLite Op as given in the documentation. Can you please check this link https://github.com/tensorflow/tensorflow/tree/r1.14/tensorflow/lite/experimental/examples/lstm/g3doc
According to this doc we can convert dynamic_rnn to tensorflow lite if we use experimental dynamic_rnn op instead of tf.nn.dynamic_rnn.
Also please check this tutorial given by tensorflow team https://github.com/tensorflow/tensorflow/blob/r1.15/tensorflow/lite/experimental/examples/lstm/TensorFlowLite_LSTM_Keras_Tutorial.ipynb. They succesfully converted to TFLite model

**Comment by haozha111** [comment]
Which version of tensorflow are you using?

could you try repeating the steps here:
https://github.com/tensorflow/tensorflow/tree/r1.14/tensorflow/lite/experimental/examples/lstm/g3doc#0-turn-on-control_flow_v2

Could you also share the converter command you use?

**Comment by tulasiram58827** [comment]
Tensorflow Version - 1.14.

For saving the model I used simple_save() function.

For freezing the graph I used :
python -m tensorflow.python.tools.freeze_graph --input_saved_model=saved_model.pb --output_node_names="output"

This is the code I used to converted the frozen graph into tflite model.

graph_def_file = pbfile_path
        input_arrays = ["inputs"]
        output_arrays = ["output"]

        converter = tf.lite.TFLiteConverter.from_frozen_graph(
          graph_def_file, input_arrays, output_arrays,input_shapes = {"inputs":[1,100,13]})
        converter.allow_custom_ops=False
        converter.target_ops = [tf.lite.OpsSet.TFLITE_BUILTINS,tf.lite.OpsSet.SELECT_TF_OPS]
        #converter.optimizations = [tf.lite.Optimize.DEFAULT]
        #converter.post_training_quantize=True
        #converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8]
        tflite_model = converter.convert()

        open("output.tflite", "wb").write(tflite_model)

**Comment by haozha111** [comment]
Thanks for the info.

I'm wondering if there are certain corner cases caused the conversion to fail. Renjie, could you help also take a look?

**Comment by renjie-liu** [comment]
Hi, you don't need to manually freeze the graph, you can just 

use converter = tf.lite.TFLiteConverter.from_session(sess, input_arrays, output_arrays,input_shapes = {"inputs":[1,100,13]})

like here
https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/experimental/examples/lstm/g3doc#3-lets-define-the-export-to-tensorflow-lite-model-function

also please note tf.lite.experimental.nn.dynamic_rnn only accepts time_major=True (and it's the default value).

**Comment by tulasiram58827** [comment]
Yeah its working I can directly convert to tensorflow lite model from session.
But what is the problem with above process I did?

**Comment by renjie-liu** [comment]
Great to know it's working. :)

If you converted from frozen graph, you need to use tf.lite.experimental.convert_op_hints_to_stubs

like here:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/python/op_hint.py#L36-L40

using 'tf.lite.TFLiteConverter.from_session' is much simpler. :)

**Comment by tulasiram58827** [comment]
Ok. Thanks for the info.

**Comment by tulasiram58827** [comment]
Hey @renjie-liu  tf.lite.TFLiteConverter.from_session doesn't have any parameter as input_shapes. Can you please check it again.

**Comment by tulasiram58827** [comment]
@renjie-liu   #33429 Can you please reopen this issue

**Comment by renjie-liu** [comment]
are you trying to use a different shape? from_session takes in tensors, which have shapes implicitly

**Comment by tulasiram58827** [comment]
Yes I am using different shapes. Because at training stage I am having shape of [None,None,13] which is not accepted by tflite converter as it is having shape of None in 2nd dimension.

**Comment by renjie-liu** [comment]
I see.

I think you have two options:

1) freeze the graph & call tf.lite.experimental.convert_op_hints_to_stubs then pass to tfliteconverter

2) refactor your model building structure, yes you have [None,None,13] for training, but export a fixed-shape graph for inference

**Comment by tulasiram58827** [comment]
How to use tf.lite.experimental.convert_op_hints_to_stubs with tf.lite.TFLiteConverter. Can you give me a sample code.
I will be trying the 2nd option mean while

**Comment by tulasiram58827** [comment]
Second option worked for me. Its fine. It will be better if you share some details regarding how to use 1st option.

**Comment by renjie-liu** [comment]
please check here: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/python/op_hint.py#L38-L41

but ignore the toco_convert (should be tfliteconverter)

**Comment by tulasiram58827** [comment]
Ok Thanks.
